### ISSUE TITLE
Improve execution count discoverability on hover/active cells

### ISSUE THREAD

**Issue by misolori** [microsoft/vscode#104680]
# Summary
In the Python user studies we found out that users had a hard time finding the execution count when they would run cells. This is because we hide the execution count when a cell is hovered or active and show the run button in place:

![Kapture 2020-08-14 at 10 14 57](https://user-images.githubusercontent.com/35271042/90275422-2388bf80-de17-11ea-88fe-0c2691cf64c6.gif)

# Explorations

## 1. Place in status bar
Closer to the execution time and near the output window

![image](https://user-images.githubusercontent.com/35271042/90276051-1e784000-de18-11ea-8ec9-f1edc8e95e0a.png)

## 2. Only show on hover of execution count
When there is no execution count, we could display the run button. When the execution count appears, then we make the play button only appear when you hover on execution count. This could introduce discoverability issues with the run button.

![Kapture 2020-08-14 at 11 28 09](https://user-images.githubusercontent.com/35271042/90281417-62237780-de21-11ea-9207-fa5abf79acda.gif)

## 3. Vertically stack
This could cause visual noise/distraction as you scroll + hover several cells at a time

![Kapture 2020-08-14 at 11 16 27](https://user-images.githubusercontent.com/35271042/90280357-a281f600-de1f-11ea-9e75-70a55190e61c.gif)

## 4. Place at the bottom of cell

This has the potential to clash with the output renderer that appears on the left of the output

![image](https://user-images.githubusercontent.com/35271042/90280575-ff7dac00-de1f-11ea-8c1e-0d6b8e4d3b9e.png)

**Comment by rebornix** [comment]
My 2 cents on above four designs.

**1**
Pro: Once users start execution, the spinner will draw your attention so you might already look at the status bar, showing the execution count there makes sense in this case

Con: When the cell is relatively large, the execution count is not that visible. Also when users run all cells in the document (usually when you reopen a document and connect to a fresh kernel), users can quickly go through the left margin of each cell to see if it finishes execution by glancing at the execution count. Moving it to the status bar makes it hard though.

**2** 
Con: Execution count is still invisible after execution if the mouse doesn't move

**3**
Con: the position shifting is distracting

**4**
This one is my favorite. One can argue that it can be invisible when the bottom of the cell is not in the viewport but IMHO not a big deal.

**Comment by greazer** [comment]
My feedback for all of these options is that they do not necessarily address the feedback from jupyter-loving DS's that we got, except possibly 3. The thing we need to solve is that the user has to KNOW that the cell has run when they click the run button with the mouse or press shift enter. Therefore, the execution count (OR some form of positive acknowledgement that the user actually pressed shift-enter or clicked the play button) needs to be visible without needing to scroll the display or move their mouse. 

3 is the only option above that solves this. As to whether it would be distracting for a significant number of users, I think that's hard to say without testing.

**Comment by claudiaregio** [comment]
My feedback:

1. We have had feedback from community asking for ability to hide the status bar. If execution count is located here they would lose that if they decide they don't want the extra padding. 

2 + 3. I think steering clear of anything hover related for the run and execution count is best. 
I believe version 2 would lead to the opposite complaint but about the run button as we already got complaints about hover in general. 
I believe version 3 would cause way too much scrolling noise. In this version if user scrolls anywhere they are guaranteed noise at least on both the left and right side of their notebook. If they scroll in the middle, they are guaranteed noise from the left, right AND the middle from the add cell/markdown icons. I personally would find this really visually heavy on the eyes having hover things pop up across the entire width of the notebook. 

4. This is one is also my favorite as it displays both run and execution count statically but if users hide the status bar, would the location just change to be in line with the last line of code? If users don't scroll down to see the spinner and just want to know if it ran, they would have to scroll down to check. Could an option be to have the execution count horizontally in line with the run button but to the left of it? A con of having it horizontally aligned with run would be if the user decides to decides to hide the cell status bar they would have no indication of a cell currently running other than execution count being missing. Considering they have been trained to look for a visual cue in original notebook experience and competitor tools this change would likely confuse them unless the spinner were to move out of status bar to the eventual location of the execution count.

**Comment by roblourens** [comment]
I also like 4, but regarding hiding the status bar, that's what I don't really understand about the concept of letting the status bar be hidden - then yeah, you don't see the information in the status bar 😁 . They would see the play button turn to a cancel button at least.

**Comment by greazer** [comment]
For the record I'm really hesitant about not displaying the execution count near the play button. This is the behavior for all notebooks that I'm aware of. We had a user indicate that they hated, HATED our current behavior, which indicates there will be some very strong opinions out there if we make too big a change from what is considered the standard. We are trying to attract Jupyter, Jupyter Lab, Nteract, etc users to VS Code. Anything that is too different from what they are used to can easily push them away even before they have a chance to understand what else VS Code Notebooks can offer.

**Comment by roblourens** [comment]
Understood. Let's try this (option 4) for a bit and see how we feel about it.

**Comment by greazer** [comment]
Sure, but I'd like to better understand why 4 is preferred to try? I'm not really seeing any reasoning that indicates and pros, only cons (hiding status bar, requiring scroll to see). Can we do an A/B test for 3 (or some variant like Claudia mentioned) and 4?

**Comment by misolori** [comment]
> I'm not really seeing any reasoning that indicates any pros

Here's a few pros:
- It's always visible and doesn't rely on hover (when not on large code blocks)
- It lines up with the status bar, which from the studies showed that nearly all participants saw it
- It's closer to the output + execution time

> only cons (the real possibility that some users will want to hide the status bar

Hiding the status bar should not have an impact on the visibility of the execution count in this version, it would just move up.

> and the need to scroll sometimes to see it

I agree that this isn't ideal but it's also an existing issue in current jupyter notebooks, like when you have a large code block and the top of the cell is offscreen:
![image](https://user-images.githubusercontent.com/35271042/90784917-9f22ba80-e2b6-11ea-9d22-2092acba07e4.png)

And regarding the comments about users hiding the status bar and not being able to see certain meta data, this is already an existing pattern in VS Code for the global status bar that some choose to hide which means they lose out on a lot of other features (git sync, problems, language selector, notification indicator, and any other extension contributions). If a user hides this they are then deciding to not see that information. Another option we could add in the "hiding status bar" discussion is to hide by default but then only show when it is executed. This provides the best of both worlds for those who don't want to see at first until additional information has been surfaced.

**Comment by roblourens** [comment]
> Hiding the status bar should not have an impact on the visibility of the execution count in this version, it would just move up.

To where? If you have a single line of code and hide the status bar, then the execution count conflicts with the run button.

**Comment by misolori** [comment]
@roblourens ah, good point. That means we'd need a min-height set on a cell then.

**Comment by greazer** [comment]
"That means we'd need a min-height set on a cell then."

Either that or place the execution count to the left of the run button where this becomes a non-issue. :) I'm really not understanding how the placement at the bottom of the cell is arguably better.

**Comment by cpoptic** [comment]
I will say @greazer has an excellent point:  a key design consideration should be a very clear indicator of "Yes this cell has executed" vs "No, this cell has not yet executed".

The Execution count placement is somewhat secondary to that.

Regarding the four design options, however:
1)  This one is not optimal for the simple reason that the cell count number is not left of the cells.  DSs want to scan through a notebook at a quick glance to find "cell number x", and this Option 1 makes it much harder to find a certain cell.  Not to mention some users could confuse that execution value with actual code in the cell, which we definitely don't want to do.

2)  I really don't like how the mouse hover over causes the execution value to disappear.  Veto this one.

3)  Not bad, but I agree with other feedback that the movement of the [1] could be distracting/annoying.

4)  I like this one except as others have pointed out, for large cells you lose the execution count value off the bottom of the screen if the cell has a bunch of lines.  This is a pretty big dealbreaker, as ideally you don't want to have to scroll down just to see which cell count you're on; it adds a tiny amount of unnecessary cognitive load.

Option 5)  What about simply vertically stacking both the [1] execution value and the triangle "Run cell" button?  Statically, unlike Option 3.  That way both the executed count value and the "Run cell" button are both always displayed, and they're both left of the cell contents, and thus less obtrusive.

So in conclusion, Option 3 (or some variant of it like my idea for Option 5) gets my vote.  But I'm just a data scientist, not a designer; I'm only voicing what I (and probably other data scientists) would value in this tool.

Btw, shout out to all the VS Code contributors; love your work.

**Comment by roblourens** [comment]
@cpoptic https://github.com/microsoft/vscode/issues/105876 is the open issue for continuing this discussion, please move your comment there, thanks
